### PR TITLE
X509 app: major cleanup of user guidance, documentation, and code structure

### LINKED ISSUE #3638: Support "copy_extensions" also with x509 CSR signing
While already supported with "openssl ca", basic signing does not support the "copy_extension" mode. Basic signing might be neccessary when the "openssl ca" magic is too much and cannot be turned off in certain usecases. Thus when using "openssl x509" instead, from each CSR, a openssl.config has to be created manually by duplicating the CSR fields before signing, which makes it even more risky and error prone than using the "copy_extensions".

It would be nice to support the existing "copy_extensions = copy" feature also in for "openssl x509". (It would be even more nice, if it would allow "... = copy:subjectAltName", but that is another story ...)

### LINKED ISSUE #6481: Extensions in certificates are not transferred to certificate requests and vice versa.
From man 1 x509:

> BUGS
> Extensions in certificates are not transferred to certificate requests and vice versa.

This bug needs to die in a fire.  It renders `x509 -x509toreq` absolutely useless.

https://rt.openssl.org/Ticket/Display.html?id=3762 suggests that it has been resolved, but I can't see any release information that suggests when.

Making a custom conf file is not a suitable workaround, unless there is an obvious single-line config that I'm missing, but `copy_extensions = copy` is only documented for CA.  I'm not a CA.  

I just don't want to have to type however many subject alternate names, basic usages, extended usages, etc. into a new config, when they already exist in the certificate.  Surely that's the whole point of `-x509toreq`?

### LINKED ISSUE #10458: Why does the x509 command not copy extension in certificate request?
Of course, I am not the first person to encounter this problem. After my search, I found that many people have raised this question.

https://github.com/openssl/openssl/issues/6481

https://github.com/Netflix/lemur/pull/1909

https://github.com/openssl/openssl/pull/9449

https://github.com/openssl/openssl/issues/3638

https://github.com/openssl/openssl/pull/9497

https://github.com/openssl/openssl/issues/6481

https://stackoverflow.com/questions/33989190/subject-alternative-name-is-not-copied-to-signed-certificate

https://stackoverflow.com/questions/6194236/openssl-version-v3-with-subject-alternative-name

https://stackoverflow.com/questions/30977264/subject-alternative-name-not-present-in-certificate

https://security.stackexchange.com/questions/150078/missing-x509-extensions-with-an-openssl-generated-certificate

https://security.stackexchange.com/questions/158166/how-to-add-altname-from-csr-file-to-crt-file-using-openssl-x509-req

https://security.stackexchange.com/questions/74345/provide-subjectaltname-to-openssl-directly-on-the-command-line

https://www.linuxquestions.org/questions/linux-software-2/get-subjectaltname-into-certificate-my-own-ca-4175479553/

https://forum.ivorde.com/openssl-certificate-authority-ca-how-to-copy-x509-extensions-from-csr-to-signed-pem-t19421.html

https://stackoverflow.com/questions/25900812/certificate-is-not-including-san-names-using-openssl

http://openssl.6102.n7.nabble.com/subjectAltName-removed-from-CSR-when-signing-td26928.html

https://mta.openssl.org/pipermail/openssl-users/2016-January/002759.html

The problem encountered by so many people is only because of a small bug here.

https://www.openssl.org/docs/man1.1.1/man1/x509.html

**BUGS
Extensions in certificates are not transferred to certificate requests and vice versa.**

Why is this problem not fixed yet? Please give me a reason.

Obviously only need to add a -copy_extensions option to solve this problem perfectly

### LINKED ISSUE #13708: The x509 and req apps should copy X.509 extensions when converting formats
When using the `X509` app to input and output certs (maybe just converting from PEM to DER) or 
when using the `req` app to input and output a CSR, 
any X.590 extensions in the input are copied.

In contract, when using the `x509` app with `-x509toreq` to convert a cert to a CSR or with `-req` for the opposite direction, or
when using the `req` app with `-x509` to convert a CSR to a cert (whereas the opposite direction is not supported),
any X.509 extensions in the input are not taken over. I find this at least counter-intuitive.
I'm currently cleaning up both apps in particular concerning their user guidance and will make clear that no extension are copied.
Yet I believe it would be better if X.509 extensions were copied also when converting between cert and CSR, at least optionally.

BTW, there is a huge overlap between these two apps, while on the other hand there are needless inconsistencies in their use. 
It would have been good, also in terms of code redundancy and maintenance, to have just one app that covers all their (pretty related) use cases.

### PR DESCRIPTION
**Pull request by DDvO**
While adding tests to #13658 I came across various shortcomings in the user guidance of the `x509` app.
Like with several other apps, there are meanwhile a lot of options and their presentation to users has become a mess.
This is likely due to various incremental additions of options having grown wild without intermediate cleanup.

I spent quite some effort getting the presentation of the options in the help output and the documentation into a logical order
and fixing descriptions where they were outdated or unclear.
During this process I inevitably found a couple of minor bugs and fixed them, 
in particular regarding unexpected combinations of options and memory leaks of the `-clrext` option.

~I've also renamed the new `-force_pubkey` option to the more consistent and less 'aggressive' name `-set_pubkey`~
and did some minor improvements to the related tests.

~I've renamed the `-signkey` option to `-key` for consistency with the `req` app
and because this better reflects that usually also the public portion is used,
retaining the old `-signkey` option for backward compatibility.~

Moreover, this adds the `-copy_extensions` option, used when transforming x509 <-> req.
Fixes #3638
Fixes #6481
Fixes #10458
Partly fixes #13708
Supersedes #9449

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by FdaSilvaYY** [comment]
typo in a commit message : `reasonalbe`

File: apps/x509.c
Comment (FdaSilvaYY) [inline apps/x509.c:798]: nit : this blank line is one line off

**Comment by DDvO** [comment]
Thanks  @FdaSilvaYY for your comments. I've fixed the reported nits.

File: apps/x509.c
Comment (InfoHunter) [inline apps/x509.c:102]: Should be `both notBefore and notAfter dates`?

**Comment by mattcaswell** [comment]
> I've also renamed the new -force_pubkey option to the more consistent and less 'aggressive' name -set_pubkey

Just back from holiday and not had time to review this PR yet, but I noticed this in your description above. `-force_pubkey` is not new. It was introduced in 43206a2d7 back in 2011, and has been available in public releases since 1.0.2

File: apps/x509.c
Comment (DDvO) [reply apps/x509.c:102]: Good point, so renamed them for consistency with the cert field names
and mentioned this in `CHANGES.md`.
I think it's not worth adding option aliases for backward compatibility, is it?

**Comment by DDvO** [comment]
> > I've also renamed the new -force_pubkey option to the more consistent and less 'aggressive' name -set_pubkey
> 
> Just back from holiday and not had time to review this PR yet, but I noticed this in your description above. `-force_pubkey` is not new. It was introduced in [43206a2](https://github.com/openssl/openssl/commit/43206a2d7cc87c959535c0f69e2aa3b364eafd6e) back in 2011, and has been available in public releases since 1.0.2

Oh right, I had confused the history of some options.
So I've reverted this change because this is not important and if done it would likely require an option alias for compatibility.

File: apps/x509.c
Comment (InfoHunter) [reply apps/x509.c:102]: > I think it's not worth adding option aliases for backward compatibility, is it?

I think it's not worth, but I don't know if this would bring more inconvenient to the users.

And I think the text of option `OPT_DATES` should be `Print both notBefore and notAfter dates` since there are no fields called Before and After in an X.509 certificate ;-)

Comment (DDvO) [reply apps/x509.c]: I've revised the help and documentation texts for all three options.

Comment (richsalz) [reply apps/x509.c]: Having three options is kinda silly, unless you make notBefore and notAfter print out machine-readable, as in the ASN1 string?

Comment (DDvO) [reply apps/x509.c]: > Having three options is kinda silly

I wonder if it's the right time to discuss such details on rather irrelevant options.
Let's focus now on getting the essence of the apps usable and well documented.

Comment (richsalz) [reply apps/x509.c]: > I wonder if it's the right time to discuss such details on rather irrelevant options.
> Let's focus now on getting the essence of the apps usable and well documented.

I wonder if now is the right time to add new names to existing options.

**Comment by DDvO** [comment]
I've renamed the `-signkey` option to `-key` for consistency with the `req` app
and because this better reflects that usually also the public portion is used,
retaining the old `-signkey` option for backward compatibility.

**Comment by DDvO** [comment]
Moreover, I've added the `-copy_extensions` option, used when transforming x509 <-> req.
Fixes #3638
Fixes #6481
Fixes #10458
Partly fixes #13708
Supersedes #9449

File: CHANGES.md
Comment (t8m) [inline CHANGES.md:234]: Please drop this change. It is useless. The options were there since ages. And you're making this inconsistent with ca command.

**Review by t8m — CHANGES_REQUESTED** [review]
Could you please squash the fixups and drop all the rename options commits? I do not think it is a good idea to rename the options since they have been there for ages.

**Comment by t8m** [comment]
The -signkey rename could be submitted as separate PR to consider. However I am still not sure it is worth it.

File: CHANGES.md
Comment (DDvO) [reply CHANGES.md:234]: > Please drop this change. It is useless. The options were there since ages. 
And you're making this inconsistent with ca command.

This change was suggested by @InfoHunter who has been with the project even longer than this option.
See the discussion above: https://github.com/openssl/openssl/pull/13711#pullrequestreview-556251087
To me, either variant is fine; does anyone else have a strong opinion on it?

**Comment by DDvO** [comment]
> The -signkey rename could be submitted as separate PR to consider. However I am still not sure it is worth it.

Before doing that, I'd like to learn if it's worth the effort of separating this change (and later merging with with conflicts).
This has been my motivation for the change:

> I've renamed the `-signkey` option to `-key` for consistency with the `req` app
and because this better reflects that usually also the public portion is used,
retaining the old `-signkey` option for backward compatibility.

What do others think about it?

**Comment by DDvO** [comment]
> Could you please squash the fixups

Done, also rebased to the latest master.

File: CHANGES.md
Comment (t8m) [reply CHANGES.md:234]: As I understand the original @InfoHunter comment it was about the help text, not about the names of the options.

**Comment by t8m** [comment]
> Before doing that, I'd like to learn if it's worth the effort of separating this change (and later merging with with conflicts).
> This has been my motivation for the change:

You could just save the commit as is for later (as it is the last one in the PR anyway) and submit it only after this PR is merged.

Adding multiple unrelated things into a single PR makes it hard to review and hard to approve if someone disagrees some of the unrelated parts.

Please note that I do not completely disagree with the -signkey rename although I would keep the new option just as an alias and would not deprecate the old one as it is probably widely used already in various scripts and it is not worth removing the compatibility even any time later. So I would not block merging such rename. But I am less keen on explicitly approving it.

File: CHANGES.md
Comment (DDvO) [reply CHANGES.md]: Argh, indeed I must have entirely misunderstood his comment!
Followed your interpretation.

**Comment by DDvO** [comment]
> > Before doing that, I'd like to learn if it's worth the effort of separating this change (and later merging with with conflicts).
> > This has been my motivation for the change:
> 
> You could just save the commit as is for later (as it is the last one in the PR anyway) and submit it only after this PR is merged.

That's a nice idea - will do that.
And I've taken that commit out of this PR as you advised.

> Adding multiple unrelated things into a single PR makes it hard to review and hard to approve if someone disagrees some of the unrelated parts.

Good point.

> Please note that I do not completely disagree with the -signkey rename although I would keep the new option just as an alias and would not deprecate the old one as it is probably widely used already in various scripts and it is not worth removing the compatibility even any time later. So I would not block merging such rename. But I am less keen on explicitly approving it.

I understand, so will adapt that PR to add an alias rather than deprecating the old option name.

**Comment by DDvO** [comment]
@t8m, your recent change requests are fulfilled.

**Comment by DDvO** [comment]
I've added a further related commit regarding `apps/x509.c`: 
Make `-x509toreq` respect `-clrext`, `-sigopt`, -and `-extfile` options
and prevent copying SKID and AKID extensions to CSRs, which make no sense there.

BTW, it's a historical accident that SKIDs and AKIDs are encoded as X.509 extensions rather than as genuine cert attributes.

**Comment by richsalz** [comment]
> BTW, it's a historical accident that SKIDs and AKIDs are encoded as X.509 extensions rather than as genuine cert attributes.

ITU didn't define it, the IETF did.  i don't think anyone would say X509 is "minimal" but it really just has only what is strictly needed ... :)

File: doc/man1/openssl-x509.pod.in
Comment (t8m) [inline doc/man1/openssl-x509.pod.in:400]: When you're editing here please s/a transforming/transforming/

Comment (DDvO) [reply doc/man1/openssl-x509.pod.in]: Done.

**Comment by DDvO** [comment]
I've further improved my most recent commit such that now the `-ext` option can be used to select which extensions to copy.
On this occasion I further simplified the overall structure of the code, removing major redundancies.

**Comment by DDvO** [comment]
BTW, my two most recent commits also fix memory leaks of the `-clrext` option, 
which surfaced in tests after I had used the same `X509_delete_ext(x, i)` pattern for `-ext`.

So I fixed both usages and added a warning to the doc of `X509v3_delete_ext()` and friends 
that their behavior (namely, not to deallocate their subject) is a bit counter-intuitive.

File: apps/x509.c
Comment (t8m) [inline apps/x509.c:189]: I do not think we need this warning. It makes the command unnecessarily noisy because the skid and akid is present in most of the regular certificates produced by CAs.

**Review by t8m — CHANGES_REQUESTED** [review]
A few nits to be fixed, otherwise looks good.

**Comment by t8m** [comment]
And yeah, we need a rebase to fix conflicts.

File: crypto/x509/x509_req.c
Comment (DDvO) [reply crypto/x509/x509_req.c]: Oops, fixed.

File: CHANGES.md
Comment (DDvO) [reply CHANGES.md]: Fixed, also a further such occurrence in the following entry.
On this occasion I've also improved the description in `CHANGES.m` on the new default SKID and AKID behavior.

File: apps/x509.c
Comment (DDvO) [reply apps/x509.c]: I've dropped it as you wrote, though it would have been an automated way of making users aware of the new feature.
I believe users would expect that extensions are copied by default (as is done when converting one cert to another),
see all the requests for the extension copy feature listed here: #10458.

File: apps/x509.c
Comment (DDvO) [reply apps/x509.c]: Done; see above.

File: doc/man1/openssl-x509.pod.in
Comment (DDvO) [reply doc/man1/openssl-x509.pod.in]: Fixed

File: doc/man1/openssl-x509.pod.in
Comment (DDvO) [reply doc/man1/openssl-x509.pod.in]: Fixed

File: apps/x509.c
Comment (DDvO) [reply apps/x509.c:189]: In my opinion it is good to have this warning.
Note that it is shown only if `-x509toreq` and `-copy_extensions copy` is given and then still it is not shown by default,
but only when also the `-ext` option is given explicitly requesting to copy `subjectKeyIdentifier` or `authorityKeyIdentifier` to the cert request, which simply makes no sense and for this reason is refused.
So this warning is only given on a nonsensical explicit request (and thus could even be made an error).

**Comment by DDvO** [comment]
> A few nits to be fixed, otherwise looks good.

> And yeah, we need a rebase to fix conflicts.

Thanks @t8m for your further round of comments; 
I've addressed them and also fixed the recent merge conflicts.

File: apps/x509.c
Comment (t8m) [inline apps/x509.c:214]: Leak of req here.

File: apps/x509.c
Comment (t8m) [reply apps/x509.c:189]: OK, I misunderstood the check slightly - if the warning is given only in case skid/akid is listed in -ext, then it is definitely ok, although it might be even better to just warn if skid/akid is listed in -ext without even checking that it is present in the certificate. So it is clear right away that trying to copy akid/skid won't work.

File: crypto/x509/x509_req.c
Comment (t8m) [inline crypto/x509/x509_req.c:242]: The indentation is still wrong for the whole if here.

**Review by t8m — CHANGES_REQUESTED** [review]
Still some nits to solve.

File: apps/x509.c
Comment (DDvO) [reply apps/x509.c]: Good spotting - fixed.

File: apps/x509.c
Comment (DDvO) [reply apps/x509.c]: All right, adapted the warning this way.

File: crypto/x509/x509_req.c
Comment (DDvO) [reply crypto/x509/x509_req.c]: Oops, fixed, and simplified the function.

**Comment by DDvO** [comment]
> Still some nits to solve.

Done.

File: crypto/x509/x509_req.c
Comment (t8m) [inline crypto/x509/x509_req.c:239]: Now looking at this particular change in broader context it looks very awkward to me to check for the NULL parameters in just this particular function but not to do it in the other similar functions around. However I am not sure it is really worth it to add the checks as we do not do that universally throughout the code - what was the motivation for this particular change?

I'd recommend dropping it from this PR and if you want to, you could submit a different PR adding the NULL checks everywhere across these functions. But perhaps it should be discussed first on openssl-project ML whether we want such checks universally.

Comment (DDvO) [reply crypto/x509/x509_req.c]: As far as i recall, my motivation for adding this check was that during the changes I did to `x509.c` I made a mistake that lead to the `req` argument to be NULL and thus to a crash.
I had learned during the CMP contribution that every public function should validate its input, and so it would certainly be an improvement to do it here. 

Yet I see your point that it appears imbalanced to not do this consistently for related functions. So following your recommendation I've dropped that change from this PR,
also because it is very bulky already anyway.

On this occasion I've squashed the remaining fixup commits.

**Review by t8m — APPROVED** [review]
This one was very hard to review so for the next time I'd suggest avoiding sideline improvements if possible. Submit them as separate PRs to keep the main topic PR more manageable.

Anyway this is really a good improvement of the code and it hopefully should not cause any serious regression.

**Comment by DDvO** [comment]
> This one was very hard to review so for the next time I'd suggest avoiding sideline improvements if possible. Submit them as separate PRs to keep the main topic PR more manageable.

Yeah, unfortunately this PR became very bulky.

It started out by improving a few options of `x509` and their documentation,
but doing so I came across more and more issues related to these changes,
and I was forced to do some further fixes also in the crypto lib to get the tests through etc.

A couple of fixes could have been done independently, though.
The later addition of the `-copy_extensions` option would better have been done in a follow-up PR after this one would have been merged, but adding the same option to the `req` app I did not resist the temptation to add it here at the same time.

On the bright side, handling this all in one PR saved us from handling some 10+ separate PRs :wink:

**Comment by DDvO** [comment]
If I had known in advance how much technical debt this PR would uncover and need to deal with I'd likely not have done it because it was a huge effort, which I ended up doing in my spare time.

I appreciate your endurance also as a reviewer.

> Anyway this is really a good improvement of the code 

Pleased to hear!

>and it hopefully should not cause any serious regression.

I hope so, too.
(And if there is one, this would mean there was a serious omission in the existing tests).

**Comment by DDvO** [comment]
Merged - thanks a lot @t8m for your effort reviewing this large PR!

**Comment by kousu** [comment]
Thank you thank you thank you

**Comment by hablutzel1** [comment]
The following problem has been apparently introduced by the current PR around . 

If you create `cert.pem`:

```
-----BEGIN CERTIFICATE-----
MIIEZTCCA02gAwIBAgIQQAF1BIMUpMghjISpDBbN3zANBgkqhkiG9w0BAQsFADA/
MSQwIgYDVQQKExtEaWdpdGFsIFNpZ25hdHVyZSBUcnVzdCBDby4xFzAVBgNVBAMT
DkRTVCBSb290IENBIFgzMB4XDTIwMTAwNzE5MjE0MFoXDTIxMDkyOTE5MjE0MFow
MjELMAkGA1UEBhMCVVMxFjAUBgNVBAoTDUxldCdzIEVuY3J5cHQxCzAJBgNVBAMT
AlIzMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAuwIVKMz2oJTTDxLs
jVWSw/iC8ZmmekKIp10mqrUrucVMsa+Oa/l1yKPXD0eUFFU1V4yeqKI5GfWCPEKp
Tm71O8Mu243AsFzzWTjn7c9p8FoLG77AlCQlh/o3cbMT5xys4Zvv2+Q7RVJFlqnB
U840yFLuta7tj95gcOKlVKu2bQ6XpUA0ayvTvGbrZjR8+muLj1cpmfgwF126cm/7
gcWt0oZYPRfH5wm78Sv3htzB2nFd1EbjzK0lwYi8YGd1ZrPxGPeiXOZT/zqItkel
/xMY6pgJdz+dU/nPAeX1pnAXFK9jpP+Zs5Od3FOnBv5IhR2haa4ldbsTzFID9e1R
oYvbFQIDAQABo4IBaDCCAWQwEgYDVR0TAQH/BAgwBgEB/wIBADAOBgNVHQ8BAf8E
BAMCAYYwSwYIKwYBBQUHAQEEPzA9MDsGCCsGAQUFBzAChi9odHRwOi8vYXBwcy5p
ZGVudHJ1c3QuY29tL3Jvb3RzL2RzdHJvb3RjYXgzLnA3YzAfBgNVHSMEGDAWgBTE
p7Gkeyxx+tvhS5B1/8QVYIWJEDBUBgNVHSAETTBLMAgGBmeBDAECATA/BgsrBgEE
AYLfEwEBATAwMC4GCCsGAQUFBwIBFiJodHRwOi8vY3BzLnJvb3QteDEubGV0c2Vu
Y3J5cHQub3JnMDwGA1UdHwQ1MDMwMaAvoC2GK2h0dHA6Ly9jcmwuaWRlbnRydXN0
LmNvbS9EU1RST09UQ0FYM0NSTC5jcmwwHQYDVR0OBBYEFBQusxe3WFbLrlAJQOYf
r52LFMLGMB0GA1UdJQQWMBQGCCsGAQUFBwMBBggrBgEFBQcDAjANBgkqhkiG9w0B
AQsFAAOCAQEA2UzgyfWEiDcx27sT4rP8i2tiEmxYt0l+PAK3qB8oYevO4C5z70kH
ejWEHx2taPDY/laBL21/WKZuNTYQHHPD5b1tXgHXbnL7KqC401dk5VvCadTQsvd8
S8MXjohyc9z9/G2948kLjmE6Flh9dDYrVYA9x2O+hEPGOaEOa1eePynBgPayvUfL
qjBstzLhWVQLGAkXXmNs+5ZnPBxzDJOLxhF2JIbeQAcH5H0tZrUlo5ZYyOqA7s9p
O5b85o3AM/OJ+CktFBQtfvBhcJVd9wvlwPsk+uyOy2HI7mNxKKgsBTt375teA2Tw
UdHkhVNcsAKX1H7GNNLOEADksd86wuoXvg==
-----END CERTIFICATE-----
```

And execute:

`openssl x509 -in cert.pem -issuer -noout`

Instead of getting the correct issuer (which is something like this):

`issuer= /O=Digital Signature Trust Co./CN=DST Root CA X3`

You get the subject: 

`issuer=C = US, O = Let's Encrypt, CN = R3`

PS: Sorry for such a naive bug report without a patch or nothing. I'm just out of time.

**Comment by DDvO** [comment]
Thanks @hablutzel1 for your bug report. 
This is now being fixed in #14340.

**Comment by DDvO** [comment]
> This is now being fixed in #14340.

Done

**Comment by hablutzel1** [comment]
Kind of off-topic here, but could you please point me to some official documentation on the OpenSSL development guidelines/practices (if it exists)?, as I have the impression that this PR includes commits that are not as granular as possible and I find it kind of dangerous as that could make commits harder to be appropriately reviewed which later could lead to the introduction of bugs... even security ones.

**Comment by DDvO** [comment]
> Kind of off-topic here, but could you please point me to some official documentation on the OpenSSL development guidelines/practices (if it exists)?

There is a contributors guide, which is mentioned whenever a PR is being submitted: 
https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md
and is part of the code base, 
but for some reason is not referenced from https://www.openssl.org/policies/.

**Comment by lzap** [comment]
For googlers, this is gonna be part of the **OpenSSL 3.0.0** release.

_(This is a longer wait than for my Playstation 5, anyway, thanks all :-)_

**Comment by stas-sbi** [comment]
I don't know if this was an intent (I guess it wasn't) but this PR made ([this](https://github.com/openssl/openssl/commit/05458fdb73dcca30edace5ad727a15d6d919e215#diff-0fb19438117c9272e0d4a812eab3dfd079ad93e5e598fda46a6f2eb65fa9ad16L668-L669) change specifically) made `-CAcreateserial` and `-CAserial` options ignored. There is no serial file created and the serial is always random now.

**Comment by giner** [comment]
I don't know if this was an intent (I guess it wasn't) but this PR made ([this](https://github.com/openssl/openssl/commit/05458fdb73dcca30edace5ad727a15d6d919e215#diff-0fb19438117c9272e0d4a812eab3dfd079ad93e5e598fda46a6f2eb65fa9ad16L668-L669) change specifically) made `-CAcreateserial` and `-CAserial` options ignored. There is no serial file created and the serial is always random now.

**Comment by DDvO** [comment]
> I don't know if this was an intent (I guess it wasn't) but this PR ([this](https://github.com/openssl/openssl/commit/05458fdb73dcca30edace5ad727a15d6d919e215#diff-0fb19438117c9272e0d4a812eab3dfd079ad93e5e598fda46a6f2eb65fa9ad16L668-L669) change specifically) made `-CAcreateserial` and `-CAserial` options ignored. There is no serial file created and the serial is always random now.

Thanks for your comment.
I've just investigated: there is indeed a regression due to a mistake of mine in commit 05458fdb73dcca30edace5ad727a15d6d919e215 in case the `-new` option is given along with `-CA`.
I'll provide a fix for this issue (and a related one).

**Comment by DDvO** [comment]
Fix is in #18373.

**Pull request closed without merging**
